File: data-structures/graphs-intro/graph.go
/*
   /----A--- \
   B         C
   |         |
   D---------E
    \___F___/
*/
package main

import "fmt"

type Vertex struct {
	value interface{}
}

func filter(slice []*Vertex, v *Vertex) []*Vertex {
	var filteredSlice []*Vertex
	for i := range slice {
		if slice[i] != v {
			filteredSlice = append(filteredSlice, slice[i])
		}
	}
	return filteredSlice
}

type Graph struct {
	adjacencyList map[*Vertex][]*Vertex
}

func (g *Graph) addVertex(v *Vertex) {
	if g.adjacencyList == nil {
		g.adjacencyList = make(map[*Vertex][]*Vertex)
	}
	g.adjacencyList[v] = []*Vertex{}
}

func (g *Graph) addEdge(v1 *Vertex, v2 *Vertex) {
	g.adjacencyList[v1] = append(g.adjacencyList[v1], v2)
	g.adjacencyList[v2] = append(g.adjacencyList[v2], v1)
}

func (g *Graph) removeVertex(v *Vertex) {
	for len(g.adjacencyList[v]) != 0 {
		poppedVertex := g.adjacencyList[v][len(g.adjacencyList[v])-1]
		g.adjacencyList[v] = g.adjacencyList[v][:len(g.adjacencyList[v])-1]
		g.removeEdge(v, poppedVertex)
	}

	delete(g.adjacencyList, v)

}

func (g *Graph) removeEdge(v1 *Vertex, v2 *Vertex) {
	g.adjacencyList[v1] = filter(g.adjacencyList[v1], v2)
	g.adjacencyList[v2] = filter(g.adjacencyList[v2], v1)

}

func (g *Graph) depthFirstRecursive(vertex *Vertex) []Vertex {
	var result []Vertex
	visited := map[*Vertex]bool{}
	var dfs func(v *Vertex)
	dfs = func(v *Vertex) {
		if v == nil {
			return
		}
		result = append(result, *v)
		visited[v] = true
		for _, neighbour := range g.adjacencyList[v] {
			if !visited[neighbour] {
				dfs(neighbour)
			}
		}

	}

	dfs(vertex)
	return result
}

func (g *Graph) breadthFirst(vertex *Vertex) []Vertex {
	var result []Vertex
	visited := map[*Vertex]bool{}
	queue := []*Vertex{vertex}
	visited[vertex] = true
	for len(queue) > 0 {
		dequedVertex := queue[0]
		result = append(result, *dequedVertex)
		queue = queue[1:]
		for _, neighbour := range g.adjacencyList[dequedVertex] {

			if !visited[neighbour] {
				visited[neighbour] = true
				queue = append(queue, neighbour)
			}
		}
	}
	return result
}

func (g *Graph) depthFirstIterative(vertex *Vertex) []Vertex {
	var result []Vertex
	visited := map[*Vertex]bool{}
	stack := []*Vertex{vertex}
	visited[vertex] = true
	for len(stack) > 0 {
		poppedVertex := stack[len(stack)-1]
		result = append(result, *poppedVertex)
		stack = stack[:len(stack)-1]
		for _, neighbour := range g.adjacencyList[poppedVertex] {

			if !visited[neighbour] {
				visited[neighbour] = true
				stack = append(stack, neighbour)
			}
		}
	}

	return result
}

func main() {
	vA := Vertex{"A"}
	vB := Vertex{"B"}
	vC := Vertex{"C"}
	vD := Vertex{"D"}
	vE := Vertex{"E"}
	vF := Vertex{"F"}

	g := Graph{}
	g.addVertex(&vA)
	g.addVertex(&vB)
	g.addVertex(&vC)
	g.addVertex(&vD)
	g.addVertex(&vE)
	g.addVertex(&vF)
	g.addEdge(&vA, &vB)
	g.addEdge(&vA, &vC)
	g.addEdge(&vB, &vD)
	g.addEdge(&vC, &vE)
	g.addEdge(&vD, &vE)
	g.addEdge(&vD, &vF)
	g.addEdge(&vE, &vF)
	fmt.Println(g.breadthFirst(&vA))
}

File: data-structures/graphs-intro/graph.js
/*
    /----A--- \ 
    B         C
    |         |
    D---------E
     \___F___/
*/

class Graph {

    constructor() {
        this.adjacencyList = {}
    }

    addVertex(name) {
        if (!this.adjacencyList[name])
            return this.adjacencyList[name] = []
    }

    addEdge(v1, v2) {
        this.adjacencyList[v1].push(v2)
        this.adjacencyList[v2].push(v1)

    }
    removeEdge(v1, v2) {
        this.adjacencyList[v1] = this.adjacencyList[v1].filter((v)=>v !== v2)
        this.adjacencyList[v2] = this.adjacencyList[v2].filter((v)=>v !== v1)

    }
    removeVertex(name) {

        while (this.adjacencyList[name].length) {

            let vertex = this.adjacencyList[name].pop()
            this.removeEdge(name, vertex)

        }

        delete this.adjacencyList[name]
    }

    depthFirsRecursive(vertex) {
        const result = [];
        const visited = {};
        const dfs = (vertex)=>{
            if (!vertex) {
                return null
            }
            result.push(vertex);
            visited[vertex] = true;

            this.adjacencyList[vertex].forEach(neighbour=>{
                if (!visited[neighbour]) {
                    return dfs(neighbour)
                }
            }
            )
        }

        dfs(vertex);
        return result;

    }
    depthFirstIterative(vertex) {
        const result = [];
        const visited = {};
        let stack = [vertex];
        visited[vertex] = true

        while (stack.length !== 0) {
            console.log(stack)
            const poppedVertex = stack.pop();
            result.push(poppedVertex);

            this.adjacencyList[poppedVertex].forEach((v)=>{
                if (!visited[v]) {
                    visited[v] = true;
                    stack.push(v);
                }
            }
            )

        }
        return result;

    }
    breadthFirst(vertex) {
        const result = [];
        const visited = {};
        let queue = [vertex];
        visited[vertex] = true

        while (queue.length !== 0) {
            console.log(queue);
            const poppedVertex = queue.shift();
            result.push(poppedVertex);
            this.adjacencyList[poppedVertex].forEach((v)=>{
                if (!visited[v]) {
                    visited[v] = true;
                    queue.push(v);
                }
            }
            )

        }
        return result;

    }
}

var g = new Graph()
g.addVertex("A");
g.addVertex("B");
g.addVertex("C");
g.addVertex("D");
g.addVertex("E");
g.addVertex("F");

g.addEdge("A", "B");
g.addEdge("A", "C");
g.addEdge("B", "D");
g.addEdge("C", "E");
g.addEdge("D", "E");
g.addEdge("D", "F");
g.addEdge("E", "F");

console.log(g.breadthFirst("A"));
